job-exec: implement ability to execute job shells

grondo opened the pull request:
This PR is still a work in progress,  but I wanted to get a PR open in its current form to ensure I'm able to address likely feedback before I go away for a bit.

This PR adds the ability of the `job-exec` service to *actually* execute job shells for a job on receipt of a `start` request from the job manager. In the current implementation, the job shell used is set from configuration (the `job-exec.job-shell` broker attribute, falling back to `/bin/true`) overridden by the `attributes.system.exec.job-shell` key of submitted jobspec.

The refactoring of the `job-exec` module to support multiple execution "implementations" did not turn out as clean as I'd hoped, and it is still a bit rough around the edges. I tried to keep nice incremental commits as I worked on the refactor, but ended up throwing it away for one large garbage commit that changes a lot of code, so I apologize for that.

Existing tests for the job-exec module were adapted so that they drive the "test" exec mode, which allows testing of exception handling etc. Therefore, tests of actual execution still need to be added (along with fixes for the things found in testing).

grondo commented:
Oh, and I should warn I haven't fully tested real execution yet, that is also on todo list :fearful: So don't be surprised if testing this out something doesn't work.

grondo commented:
Ah, I also just realized there are some other sharness tests that need to be updated to use the "test" exec implementation. So some sharness tests will fail in travis...

garlick commented:
I like the implementation abstraction.  If we end up with something special for compute nodes that don't run full brokers, that will slot in nicely.

grondo commented:
>  If we end up with something special for compute nodes that don't run full brokers, that will slot in nicely.

I was also thinking we might need a slightly different exec implementation when we need to exec the IMP (thus the abstraction of bulk execution interface), however perhaps when we get to that point we'll see a simpler way.

chu11 commented on src/modules/job-exec/bulk-exec.c:
```diff
@@ -0,0 +1,400 @@
+    flux_watcher_stop (exec->check);
+}
+
+static int exec_start_cmds (struct bulk_exec *exec, int max)
```
noticed `max` is only called with `-1` set and isn't configurable via something in the API.  Perhaps don't need this?  Or later allow the user to set some type of "flow-control"-ish kinda setting?

grondo replied:
yes, this was for later. Likely not needed for the near future, but I need to get this PR done before I go so unfortunately didn't want to focus on that right now.

Note that also there is placeholder code for partial release which is unused. If it is ok, I'd like that to stay in as well.

grondo replied:
Actually I ended up making the "max rexec per loop" a configurable parameter in bulk_exec, set to 1 by default. This will prevent the job-exec module loop from "pausing" to call `flux_subprocess_rexec()` for large-ish jobs and more fairly handle other exec module traffic (however, it probably won't be an issue either way)

At least now we could tune it (and I don't have to dump the existing code)

chu11 commented on src/modules/job-exec/bulk-exec.c:
```diff
@@ -0,0 +1,400 @@
+    }
+    if (!exec->exit_batch_timer) {
+        flux_reactor_t *r = flux_get_reactor (exec->h);
+        exec->exit_batch_timer =
```
i assume this timer is so you don't call a callback within a callback?  Probably just needs a comment to that affect.

grondo replied:
No it is to batch up multiple job-shell exits that occur within a period to (eventually) support partial release. I'll try to add some comments.

grondo replied:
Ok, latest push should address this comment. Thanks!

grondo commented:
Pushed some incremental updates and experimental tests. Nothing to see here yet -- still have to address @chu11's comments.

grondo commented:
Ok, I think this is getting close to the amount of work I'll be able to get ready by tomorrow. I'll continue to do more testing and add more sharness tests (If I can think of any more)

After @chu11 at least has had another review, I'll squash this down to remove the incremental development and we could consider a merge.

One important note: currently the job-exec module kills all job shells on *any* job exception. This includes `flux job cancel`, so that is not the long term plan. However, for now we don't have the job signaling mechanism fully spec'd out, and also who initiates the signal to clean up on fatal errors, etc. Once we have that down we can re-architect exception handling here.

grondo commented:
Other todo items in `job-exec` (off the top of my head). I'll try to get these into issues.



 * Epilog: no post-job-shell epilog is supported yet. Probably only would be used by the system instance. Parity with wrexec epilog could be handled in the job shell.
 * Partial release; This should be pretty easy to accomplish given the new structure, but perhaps isn't necessary until we have epilog or other cleanup support (we don't want a slow epilog to tie up *all* resources of a job)
 * Job shell output: currently any job shell output is directed to the flux log. We'll probably want to tie this to the same output service used by the job shell so it can optionally be "tailed" and saved to the job record.

grondo commented:
Just pushed a fix for a small race early in the `t2402-job-exec-dummy.t` tests.

I've also found a leak in the `exec_kill` handling. The continuation callback for the future returned by `bulk_exec_kill()` is never called. I've probably done something dumb in the composite "wait all" future handling, but I'm still looking.

Once I fix that and hopefully #2203 goes in, I'll rebase and squash this PR as a candidate for merge.

grondo commented:
Ok, I've squashed this PR and rebased on current master.

There is still one issue with this implementation when a running job is killed due to an exception. The continuation for `bulk_exec_kill()` never runs, so the returned future is leaked. There is probably a bug in how the composite future is being set up.

I'm not sure I'll find a fix before I leave, though, so I'm posting this PR as-is as a candidate for another review and possible merge.

grondo commented:
Ok, found the leak and force-pushing the fixes.

garlick commented:
Just skimmed through the changes and LGTM!  This will be a solid base to rebase and retool #2196 on.  Thanks for all the hard work competing with packing for vacation!

grondo commented:
Thanks!
